add option to create domain using regional endpoint configuration

MentalBrake opened the pull request:
following this issue: https://github.com/amplify-education/serverless-domain-manager/issues/46

I added support for regional endpoint configuration

aoskotsky-amplify commented on index.js:
```diff
@@ -82,6 +84,14 @@ class ServerlessCustomDomain {
+  }
+
+  setAcmRegion(region) {
+    this.acmRegion = region || 'us-east-1';
```
Should this check that the `endpointType` is `REGIONAL`?

MentalBrake replied:
I though about this but didn't come to a conclusion yet, if only endpointType REGIONAL is provided without acm region, or some other region than us-east-1 we can throw an exception. 

What I have a problem with is if endpointType EDGE is provided with some acm region. We can choose to just ignore it, but that doesn't quite feel right because we now have a parameter that has no impact whatsoever on the domain and maybe the user isn't aware of it. maybe we could add a warning in this case letting the user know...

For now, I covered it with a short explanation in the documentation regarding the relationship between the endpoint type and the acm region.

wdyt

aoskotsky-amplify commented on index.js:
```diff
@@ -82,6 +84,14 @@ class ServerlessCustomDomain {
   }
 
+  setEndpointType(endpointType) {
+    this.endpointType = endpointType || 'EDGE';
```
Maybe should validate the value of `endpointType` somehow to make sure its either `EDGE` or `REGIONAL`

aoskotsky-amplify commented:
Thanks for the PR! Just have a few comments. Would also be great if you can add some unit tests.

MentalBrake commented:
Will add some unit tests sometime later today

majones-amplify commented on index.js:
```diff
@@ -82,6 +86,15 @@ class ServerlessCustomDomain {
 
+  setEndpointType(endpointType = 'EDGE') {
+    if (!SUPPORTED_ENDPOINT_TYPES.includes(endpointType)) throw new Error(`${endpointType} is not supported endpointType, use EDGE or REGIONAL.`);
+    this.endpointType = endpointType || 'EDGE';
```
Can we drop the `|| 'EDGE'` since there's a default parameter?

MentalBrake replied:
we definitely should, I was sure I removed it...

majones-amplify commented on index.js:
```diff
@@ -82,6 +86,15 @@ class ServerlessCustomDomain {
   }
 
+  setEndpointType(endpointType = 'EDGE') {
+    if (!SUPPORTED_ENDPOINT_TYPES.includes(endpointType)) throw new Error(`${endpointType} is not supported endpointType, use EDGE or REGIONAL.`);
```
Do the endpoint types need to be capitalized? Serverless seems to standardize on lowercase wherever possible

majones-amplify commented:
I'm unable to run `create_domain` for a regional endpoint; each time it results in a:
```
  Error --------------------------------------------------
 
  TooManyRequestsException: Too Many Requests api-east.domain.com was not created.
 
     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
 
  Stack Trace --------------------------------------------
 
Error: TooManyRequestsException: Too Many Requests api-east.domain.com was not created.
    at Promise.all.then.then.catch (/api/node_modules/serverless-domain-manager/index.js:65:15)
    at <anonymous>
    at process._tickDomainCallback (internal/process/next_tick.js:228:7)
From previous event:
    at PluginManager.invoke (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:242:22)
    at PluginManager.run (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:261:17)
    at variables.populateService.then (/usr/local/lib/node_modules/serverless/lib/Serverless.js:99:33)
    at runCallback (timers.js:789:20)
    at tryOnImmediate (timers.js:751:5)
    at processImmediate [as _immediateCallback] (timers.js:722:5)
From previous event:
    at Serverless.run (/usr/local/lib/node_modules/serverless/lib/Serverless.js:86:74)
    at serverless.init.then (/usr/local/lib/node_modules/serverless/bin/serverless:39:50)
    at <anonymous>
```

Is this something you've observed?

MentalBrake commented:
Just checked it a few times with regional endpoint type and it works fine.

I get this error from time to time, also in aws console if I don't wait a little between requests (so annoying) but I assume that you waited a little and still got it.

make sure you run the most updated code.
what parameters do you use?
node version?
aws-sdk version?
do you get the same error at the same time in aws console?

majones-amplify commented on index.js:
```diff
@@ -194,8 +210,8 @@ class ServerlessCustomDomain {
-   *  Adds the domain name and distribution domain name to the CloudFormation outputs
-   */
+ *  Adds the domain name and distribution domain name to the CloudFormation outputs
+ */
```
Not sure why the linter didn't catch this -- could we restore the indents here?

majones-amplify commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "serverless-domain-manager",
-  "version": "1.1.19",
+  "version": "1.1.20",
```
I think this is more of a minor release, not a patch fix, so maybe `1.2.0`?

majones-amplify commented:
Sorry to bother you with that, that error was entirely my fault and unrelated to this PR, I guess I'd never created many domains in rapid succession before. Couple more things but this looks pretty good!

majones-amplify commented on index.js:
```diff
@@ -304,10 +304,12 @@ class ServerlessCustomDomain {
-    return createDomain.then(data => data.distributionDomainName || data.regionalDomainName);
+    /* This will return the distributionDomainName (used in changeResourceRecordSet)
+      if the domain name already exists, the distribution domain name will be returned */
+    return this.getDomain().then(data => data.distributionDomainName).catch(() => {
```
I think the `|| data.regionalDomainName` needs to be added to this call as well, ya?

MentalBrake replied:
yep, adding it now

aoskotsky-amplify commented on .vscode/launch.json:
```diff
@@ -0,0 +1,14 @@
+{
```
This file should probably be git ignored

brettdh commented:
@aoskotsky-amplify could you give this another look when you have a chance? I may need this functionality in the near future and so I'd love to see this merged soon.

alexdebrie commented:
@MentalBrake I love this feature! Thanks for adding it.

I'm having trouble getting a Route53 record when using the regional endpoints. I think it's because the [cloudfrontHostedZoneId](https://github.com/MentalBrake/serverless-domain-manager/blob/1ad13154d6e49b50113dd9bef86caa37c68eba0d/index.js#L15) is hard-coded, which was valid for EDGE endpoints but not for REGIONAL endpoints.

The proper HostedZoneId should be available from the [`getdomain()`](https://github.com/MentalBrake/serverless-domain-manager/blob/1ad13154d6e49b50113dd9bef86caa37c68eba0d/index.js#L493) function with the key `regionalHostedZoneId`. I tried implementing but my async/NodeJS is such a mess that it'd probably just get ripped out anyway 😄 .

Does that make sense, or am I doing something wrong?

MentalBrake commented:
@alexdebrie Thanks and thanks for pointing that out,
I looked into it, and it appears that the aws sdk version that is being used in this project does not contain the regionalHostedZoneId in the domain create and get function responses.
I updated aws version, made some changes and now it works fine, no hardcoded hostedZoneId anymore.

I also deleted the whole code that was dealing with getting the hostedZoneId as it no longer necessary, but I may be wrong. good thing god blessed us with pull requests :P

MentalBrake commented:
@alexdebrie You are absolutely right, the hostedZoneIds are different. Sorry about that.
I fixed it and made sure it works this time for both regional and edge. Can you see please if it works now?

alexdebrie commented:
@MentalBrake Thanks! I just tested it quickly, and it appeared to work. I will do further testing tomorrow or the next day.

One note: I had set `hostedZoneId` in my `serverless.yml` from when I was testing a prior commit. It caused a failure just now when deploying with the `regional` endpoint, which I thought was from the code as I had forgotten I kept it in the `serverless.yml`. Is there a reason for adding the `hostedZoneId` param in the custom block for this plugin? If so, we might want to add some safeguards around it. If not, let's pull it out 😄 .

MentalBrake commented:
@alexdebrie I'm not sure about this one, this is the pr where they added this feature https://github.com/amplify-education/serverless-domain-manager/pull/66.
@rodrigobotti, @aoskotsky-amplify can any of you say if after this pr setting the hosted zone id manually is still a requirement?

alexdebrie commented:
@MentalBrake Ahh, that makes sense. Let's leave that option in then.

I think we want to avoid passing `this.hostedZoneId` to the `DomainResponse` constructor. The HostedZoneId in the DomainResponse object should be for the APIGW domain, which is a CloudFront distribution. The HostedZoneId in the `custom` block of `serverless.yml` would be the HostedZoneId of the Route53 domain that I own and am able to add record sets to.

MentalBrake commented:
removed custom hosted zone id form DomainResponse

alexdebrie commented:
@MentalBrake This worked for me today 🎉 Thanks for your work on it.

To the maintainers (@aoskotsky-amplify @majones-amplify ), is there anything we can do to make this merge-able?

jconstance-amplify commented:
Hmm. I seem to be running into some issues when I try to get this working locally:

The Error:
```
  Error --------------------------------------------------

  Error: 'REDACTED.amplify.com' was not created in API Gateway.
BadRequestException: Invalid certificate ARN: arn:aws:acm:us-east-1:646102706174:certificate/2405ea9f-280a-4caa-9147-51e00fff8bfb. Certificate must be in 'us-west-2'.

     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.

  Stack Trace --------------------------------------------

Error: Error: 'domain-manager-quotes-regional2.aws.wgen.net' was not created in API Gateway.
BadRequestException: Invalid certificate ARN: arn:aws:acm:us-east-1:646102706174:certificate/2405ea9f-280a-4caa-9147-51e00fff8bfb. Certificate must be in 'us-west-2'.
    at createDomainName.catch (/Users/jconstance/Documents/workspace/serverless-domain-manager/index.js:68:15)
    at <anonymous>
    at process._tickDomainCallback (internal/process/next_tick.js:228:7)
From previous event:
    at PluginManager.invoke (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:368:22)
    at PluginManager.run (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:399:17)
    at variables.populateService.then (/usr/local/lib/node_modules/serverless/lib/Serverless.js:102:33)
    at runCallback (timers.js:800:20)
    at tryOnImmediate (timers.js:762:5)
    at processImmediate [as _immediateCallback] (timers.js:733:5)
From previous event:
    at Serverless.run (/usr/local/lib/node_modules/serverless/lib/Serverless.js:89:74)
    at serverless.init.then (/usr/local/lib/node_modules/serverless/bin/serverless:42:50)
    at <anonymous>
```

And the relevant config:
```
custom:
    customDomain:
        basePath: ''
        domainName: "REDACTED-${opt:stage}.amplfy.com"
        stage: "${opt:stage}"
        endpointType: "regional"
        certificateRegion: "us-east-1"
```

Any thoughts on where I might be going wrong with this?

alexdebrie commented:
@jconstance-amplify What's the `region` for your service? Is it `us-west-2`, based on this message (scroll to the very end):

```
BadRequestException: Invalid certificate ARN: arn:aws:acm:us-east-1:646102706174:certificate/2405ea9f-280a-4caa-9147-51e00fff8bfb. Certificate must be in 'us-west-2'
```

If so, you'll need to change `certificateRegion` in your `custom` block:

```yml
custom:
    customDomain:
        basePath: ''
        domainName: "REDACTED-${opt:stage}.amplfy.com"
        stage: "${opt:stage}"
        endpointType: "regional"
        certificateRegion: "us-west-2"
```

Also, I'm about to run to a meeting but want to circle back on this. I wonder if we can remove `certificateRegion` and just do the following psuedo-code:

If endpointType == "edge", certificateRegion = us-east-1.

If endpointType == "region", certificateRegion = this.provider.getRegion();

cc @MentalBrake

jconstance-amplify commented:
@alexdebrie Ah, okay. That worked.

Ya, it seems like we could definitely simplify things by removing the `certificateRegion` configuration option.

alexdebrie commented:
@jconstance-amplify Yep, I think that works.

@MentalBrake Can you do the following:

1. Remove `this.serverless.service.custom.customDomain.certificateRegion` from [this line](https://github.com/MentalBrake/serverless-domain-manager/blob/633ce372f689db9dab4454ed5f44ce93a56e6a15/index.js#L56).

2. Change the [`setAcmRegion()` method](https://github.com/MentalBrake/serverless-domain-manager/blob/633ce372f689db9dab4454ed5f44ce93a56e6a15/index.js#L109-L111) to:

```javascript
  setAcmRegion() {
    if (this.endpointType == "EDGE") {
      this.acmRegion = this.serverless.getProvider('aws').getRegion()
    } else {
      this.acmRegion = "us-east-1"
  }
```

MentalBrake commented:
@jconstance-amplify I made the certificate region initialization automatic, but the build in travis seems to be stuck...

jconstance-amplify commented:
@MentalBrake I cancelled and restarted the build and looks like its going. Probably related to the recent Travis outage [here](https://www.traviscistatus.com/incidents/c1s4dlyxd4lf).

On another note, I played around with the most recent changes, they seem to work okay locally. One thing I noticed is that it doesn't look like I can convert an existing custom domain from edge to regional, or vice versa. Is that something that was intended to be supported in this PR?

MentalBrake commented:
@jconstance-amplify you are right, the domain name should be updated with any changes presented in serverless.yml, but from what I've seen this is not the case in the current version, you either get the domain or create it if it doesn't exist, though route53 is updated in any case, the domain name isn't.

We should definitely add this functionality but I don't think it's in the scope of this pr. of course if the update functionality was already there, I would update the endpoint type too.

you can confirm it by trying to change the certificate name in serverless.yml using master branch

jconstance-amplify commented:
@MentalBrake Fair enough. I'll add a known issue regarding the actual custom domains not being updated, and then we can merge this.

jconstance-amplify commented:
@MentalBrake @alexdebrie Thanks to both of you for your help with this issue! :D

alexdebrie commented:
Thanks, @jconstance-amplify and @MentalBrake !
